# built-in imports
import json
import pickle
from datetime import datetime
from glob import glob
from itertools import (product, repeat)
from multiprocessing import Pool
from os import makedirs
from os.path import (basename, dirname, exists, join, splitext)

# external libraries
import snekmer as skm
import numpy as np
from pandas import (DataFrame, read_csv, read_json)
from Bio import SeqIO
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier


# collect all fasta-like files, unzipped filenames, and basenames
input_files = glob(join("input", "*"))
zipped = [fa for fa in input_files if fa.endswith('.gz')]
unzipped = [fa.rstrip('.gz') for fa, ext
            in product(input_files, config['input']['file_extensions'])
            if fa.rstrip('.gz').endswith(f".{ext}")]

# map extensions to basename (basename.ext.gz -> {basename: ext})
uz_map = {skm.utils.split_file_ext(f)[0]: skm.utils.split_file_ext(f)[1] for f in zipped}
fa_map = {skm.utils.split_file_ext(f)[0]: skm.utils.split_file_ext(f)[1] for f in unzipped}
UZS = list(uz_map.keys())
FAS = list(fa_map.keys())
NON_BGS, BGS = FAS, []

# parse any background files
if str(config['input']['background']) != "None":
    bg_files = config['input']['background']
    if isinstance(bg_files, str):
        bg_files = [bg_files]
    bg_files = [skm.utils.split_file_ext(basename(f))[0] for flist in
                [glob(join("input", bg)) for bg in bg_files]
                for f in flist]
    NON_BGS, BGS = [f for f in FAS if f not in bg_files], bg_files

print(NON_BGS, BGS)


# validity check for alphabets
if (config['map_function'] not in skm.alphabet.ALPHABETS.keys()) and (str(config['map_function']) != "None"):  # and config['map_function'] != 'custom':
    raise ValueError("Invalid alphabet specified; alphabet must be a"
                     " string in the form 'reduced_alphabet_n' where"
                     " n is an integer between 0 and"
                     f" {len(skm.alphabet.ALPHABETS)}.")
# elif config['map_function'] == 'custom':


# define output directory (helpful for multiple runs)
if config['output']['nested_dir']:
    out_dir = join("output", config['map_function'], f"k-{config['k']:02}")
else:
    out_dir = "output"


# import subworkflow for unzipping any zipped input files
# subworkflow unzip:
#     # workdir:
#     #     join("tune")
#     snakefile:
#         join("rules", "unzip.smk")
#     configfile:
#         join("rules", "unzip_config.yaml")


# define main workflow
rule all:
    input:
        expand(join("input", '{uz}'), uz=UZS),
        expand(join(out_dir, "features", "{nb}", "{fa}.json"), nb=NON_BGS, fa=FAS),
        # expand(join(out_dir), "score", "{nb}.csv", nb=NON_BGS),
        expand(join(out_dir, "model", "{nb}.pkl"), nb=NON_BGS)
        # expand(join(out_dir, "score", "{fa}.json"), fa=FAS)


# [in-progress] kmer walk
if config['walk']:
    rule perform_kmer_walk:
        input:
            fasta=get_fasta_files
        output:
            # need to fix code to properly generate an output...
        run:
            skm.walk.kmer_walk(input.fasta)


# if any files are gzip zipped, unzip them
if len(UZS) > 0:
    rule unzip:
        input:
            lambda wildcards: join("input", f"{wildcards.uz}.{uz_map[wildcards.uz]}.gz")
            # lambda wildcards: unzip(join("input", ))
        output:
            join("input", "{uz}.{uzext}")
        params:
            outdir=join("input", 'zipped')
        run:

    # shell:
    #     "mkdir {params.outdir} && cp {input} {params.outdir} && gunzip -c {input} > {output}"
            if wildcards.sample.endswith('.fastq'):
                shell("echo gzip {input}")
                shell("echo mv {input}.gz {params.outdir}")
            else:
                shell("mv {input} {params.outdir}")


# read and process parameters from config
rule preprocess:
    input:
        fasta=lambda wildcards: join("input", f"{wildcards.nb}.{fa_map[wildcards.nb]}"),
    output:
        data=join(out_dir, "processed", "{nb}.json"),
        desc=join(out_dir, "processed", "{nb}_description.csv")
    log:
        join(out_dir, "processed", "log", "{nb}.log")
    run:
        # log step initialization
        start_time = datetime.now()

        # read fasta file
        seq_list, id_list = skm.io.read_fasta(input.fasta)

        # if random alphabet specified, implement randomization
        if config['randomize_alphabet']:
            rand_alphabet = skm.transform.randomize_alphabet(config['input']['map_function'])
            map_function = [residues, map_name, rand_alphabet]
        else:
            map_function = config['map_function']
            if str(map_function) == "None":
                map_function = None

        # if no feature set is specified, define feature space
        if not config['input']['feature_set']:
            # prefilter fasta to cut down on the size of feature set
            filter_dict = skm.features.define_feature_space(
                {k: v for k, v in zip(id_list, seq_list)},
                config['k'],
                map_function=map_function,
                start=config['start'],
                end=config['end'],
                min_rep_thresh=config['min_rep_thresh'],
                verbose=config['output']['verbose'],
                log_file=log[0],
                processes=config['processes']
                )
            filter_list = list(filter_dict.keys())
            assert len(filter_list) > 0, "Invalid feature space; terminating."
        else:
            # read in list of ids to use from file; NO FORMAT CHECK
            filter_list = []
            with open(config['input']['feature_set'], "r") as f:
                for line in f.readlines():
                    filter_list.append(line.split()[0])

        # optional indexfile with IDs of good feature output examples
        if config['input']['example_index_file']:
            example_index = skm.io.read_example_index(
                config['input']['example_index_file']
                )
        else:
            example_index = {}

        # loop thru seqs, apply input params to preprocess seq list
        seen = []  # filter duplicates
        save_data = dict()

        # define recursive and nonrecursive saving patterns for params
        recursive = ['sequences', 'ids', 'residues']
        nonrecursive = ['map_function', 'k', 'example_index', 'filter_list']
        all_dsets = recursive + nonrecursive

        for i in range(len(seq_list)):
            seq = seq_list[i]
            sid = id_list[i]

            # ignore duplicate ids
            if config['output']['filter_duplicates'] and sid in seen:
                continue
            seen.append(sid)

            seqs = [seq]
            sids = [sid]

            # shuffle the N-terminal sequence n times
            if config['output']['shuffle_n']:
                example_index[id] = 1.0
                scid_list, scramble_list, example_index = skm.transform.scramble_sequence(
                    sid, seq[:30], n=config['output']['shuffle_n'],
                    example_index=example_index
                    )
                seqs += scramble_list
                sids += scid_list

                # include shuffled sequences in output
                if config['output']['shuffle_sequences']:
                    filename = join(out_dir, 'shuffled',
                                    wildcards.fa, "%s_shuffled.fasta" % sid)
                    if not exists(dirname(filename)):
                        makedirs(dirname(filename))
                    with open(filename, "w") as f:
                        for i in range(len(sids)):
                            f.write(">%s\n%s\n" % (sids[i], seqs[i]))

            # run SIEVE on the wt and each shuffled sequence
            if config['output']['n_terminal_file']:
                sids_n, seqs_n = skm.transform.make_n_terminal_fusions(
                    sid, config['output']['n_terminal_file']
                    )
                seqs += seqs_n
                sids += sids_n
            residues = None
            if config['nucleotide']:
                residues = "ACGT"

            # populate dictionary for json save file
            to_save = [seqs, sids, residues]
            save_label = recursive
            for dset, label in zip(to_save, save_label):
                if label in save_data.keys() and save_data[label] is not None:
                    save_data[label] = save_data[label] + dset
                else:
                    save_data[label] = dset

        # save variables not generated in loop
        for dset, label in zip(
            [map_function, config['k'], example_index, filter_list],
            nonrecursive
        ):
            save_data[label] = dset

        # save all parameters into json
        with open(output.data, 'w') as f:
            json.dump(save_data, f)

        # read and save fasta descriptions into dataframe
        try:
            desc = skm.utils.parse_fasta_description(input.fasta)
            desc.to_csv(output.desc)
        except AttributeError:  # if no description exists > empty df
            DataFrame([]).to_csv(output.desc)

        # record script runtime
        end_time = datetime.now()
        with open(log[0], 'a') as f:
            f.write(f"start time:\t{start_time}\n")
            f.write(f"end time:\t{end_time}\n")
            f.write(f"total time:\t{skm.utils.format_timedelta(end_time - start_time)}")


rule generate_kmers:
    input:
        params=rules.preprocess.output.data
    output:
        labels=join(out_dir, "labels", "{nb}.txt")
    log:
        join(out_dir, "labels", "log", "{nb}.log")
    run:
        start_time = datetime.now()

        # read processed features
        with open(input.params, 'r') as f:
            params = json.load(f)

        # generate labels only
        labels = skm.transform.generate_labels(
            config['k'],
            map_function=params['map_function'],
            # residues=params['residues'],
            filter_list=params['filter_list']
            )
        if config['output']['format'] == "simple":
            skm.features.output_features(
                output.labels, "matrix", labels=labels
                )

        # record script runtime
        end_time = datetime.now()
        with open(log[0], 'a') as f:
            f.write(f"start time:\t{start_time}\n")
            f.write(f"end time:\t{end_time}\n")
            f.write(f"total time:\t{skm.utils.format_timedelta(end_time - start_time)}")


rule standardize_kmers:
    input:
        kmers=rules.generate_kmers.output.labels,
        params=rules.preprocess.output.data,
        fastas=unzipped
    log:
        join(out_dir, "features", "log", "{nb}.log")
    output:
        files=expand(join(out_dir, "features", "{{nb}}", "{fa}.json"),
                     fa=FAS)
    run:
        start_time = datetime.now()

        # get kmers for this particular set of sequences
        kmers = skm.io.read_output_kmers(input.kmers)

        # read processed features
        with open(input.params, 'r') as f:
            params = json.load(f)

        # sort i/o lists to match wildcard order
        fastas = sorted(input.fastas)
        # print(fastas)
        outfiles = sorted(output.files)

        # revectorize based on full kmer list
        for i, fa in enumerate(fastas):
            results = {'seq_id': [], 'vector': []}
            seq_list, id_list = skm.io.read_fasta(fa)
            for seq, sid in zip(seq_list, id_list):
                results['seq_id'] += [sid]
                results['vector'] += [skm.transform.vectorize_string(
                    seq, config['k'], params['map_function'],
                    start=config['start'],
                    end=config['end'],
                    filter_list=kmers,  # params['filter_list'],
                    verbose=False,  # way too noisy for batch process
                    log_file=log[0]
                )]

            with open(outfiles[i], 'w') as f:
                json.dump(results, f)

        # record script runtime
        end_time = datetime.now()
        with open(log[0], 'a') as f:
            f.write(f"start time:\t{start_time}\n")
            f.write(f"end time:\t{end_time}\n")
            f.write(f"total time:\t{skm.utils.format_timedelta(end_time - start_time)}")


rule score_features:
    input:
        kmers=rules.generate_kmers.output.labels,
        files=rules.standardize_kmers.output.files
    output:
        df=join(out_dir, "features", "{nb}.csv"),
        scores=join(out_dir, "score", "{nb}.csv")
    log:
        join(out_dir, "score", "log", "{nb}.log")
    run:
        start_time = datetime.now()
        with open(log[0], 'a') as f:
            f.write(f"start time:\t{start_time}\n")

        # get kmers for this particular set of sequences
        kmers = skm.io.read_output_kmers(input.kmers)

        # parse all data and label background files
        label = config['score']['lname']
        data = skm.io.vecfiles_to_df(
            input.files, labels=config['score']['labels'], label_name=label
            )
        data['background'] = [skm.utils.split_file_ext(f)[0] in BGS for f in data['filename']]

        timepoint = datetime.now()
        with open(log[0], 'a') as f:
            f.write(f"vecfiles_to_df time:\t{skm.utils.format_timedelta(timepoint - start_time)}\n")

        # parse family names and only add if some are valid
        families = [skm.utils.get_family(fn) for fn in data['filename']]
        if any(families):
            label = 'family'
            data[label] = families

        # define feature matrix of kmer vectors not from background set
        non_bg = data[~data['background']]
        feature_matrix = skm.score.to_feature_matrix(non_bg['vector'].values)

        # compute class probabilities
        labels = non_bg[label].values
        class_probabilities = skm.score.feature_class_probabilities(
            feature_matrix.T, labels, kmers=kmers
            )#.drop(columns=['count'])

        # assign family probability scores
        for fam in non_bg[label].unique():
            scores = class_probabilities[
                class_probabilities['label'] == fam
            ]['score'].values

            # normalize by sum of all positive scores
            norm = np.sum([s for s in scores if s > 0])

            # include background sequences for score generation
            full_feature_matrix = skm.score.to_feature_matrix(data['vector'].values)
            total_scores = skm.score.apply_feature_probabilities(
                full_feature_matrix, scores, scaler=config['score']['scaler'],
                **config['score']['scaler_kwargs']
            )

            # assign percent score based on max positive score
            data[f"{fam}_score"] = total_scores / norm * 100

        new_timepoint = datetime.now()
        with open(log[0], 'a') as f:
            f.write(
                "class_probabilities time:\t"
                f"{skm.utils.format_timedelta(new_timepoint - timepoint)}\n"
                )

        # [IN PROGRESS] compute clusters
        clusters = skm.score.cluster_feature_matrix(full_feature_matrix)
        data['cluster'] = clusters

        # save all files to respective outputs
        delete_cols = ['vec', 'vector']
        for col in delete_cols:
            if col in data.columns:
                data = data.drop(columns=col)
            if col in class_probabilities.columns:
                class_probabilities = class_probabilities.drop(columns=col)
        data.to_csv(output.df, index=False)
        class_probabilities.to_csv(output.scores, index=False)

        # record script runtime
        end_time = datetime.now()
        with open(log[0], 'a') as f:
            f.write(f"total time:\t{skm.utils.format_timedelta(end_time - start_time)}")


rule build_model:
    input:
        files=rules.standardize_kmers.output.files,
        scores=rules.score_features.output.scores
    output:
        model=join(out_dir, "model", "{nb}.pkl"),
        results=join(out_dir, "model", "{nb}.csv")
    run:
        data = skm.model.format_data_df(input.files)
        scores = read_csv(input.scores)
        family = skm.utils.get_family(input.scores)

        # define x, y matrices
        X, y, Y = np.asarray([arr for arr in data['vec'].values]), data[family].values, data['family'].values
        Xb = (X > 0) * 1  # binary presence/absence

        # restrict kmer vector by score
        if config['model']['use_score']:
            scaler = skm.model.KmerScaler(n=config['model']['n'])
            score_list = scores[scores['label'] == family]['score'].values[0]
            scaler.fit(score_list)
            Xb = scaler.transform(Xb)

        Xb_train, Xb_test, y_train, y_test = train_test_split(
            Xb, y, test_size=0.25, stratify=y
            )

        # train and save model
        clf = skm.model.KmerModel(model=config['model']['type'], scaler=None)  # scaler currently separate from model
        print(clf.model)
        clf.fit(Xb_train, y_train)
        print(clf.score(Xb_test, y_test))

        # save cv results and model
        DataFrame(clf.search.cv_results_).to_csv(output.results, index=False)
        with open(output.model, 'wb') as f:
            pickle.dump(clf, f, protocol=pickle.HIGHEST_PROTOCOL)
